File: hsg12-tinh/2023-2024-vinhphuc/squery/squery.cpp
#include <iostream>
#include <vector>

#define input_file "squery.inp"
#define output_file "squery.out"

using namespace std;

int n, q;
vector<int> A;
vector<int> result;

// Hàm dùng để đếm số phần tử là ước hoặc bội của d
int count_dm(int l, int r, int d)
{
    int cnt = 0;

    // Duyệt mảng A từ vị trí l đến vị trí r
    for (int i = l; i < r +  1; ++i)
    {
        // Xét A[i] là ước hoặc bội của d
        if (A[i] % d == 0 || d % A[i] == 0)
        {
            cnt += 1;
        }
    }

    return cnt;
}


void input()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    freopen(input_file, "r", stdin);

    cin >> n >> q;

    // Cấp phát bộ nhớ cho n + 1 phần tử (vì mảng A đánh chỉ số từ 1)
    A.resize(n + 1);

    for (int i = 1; i < n + 1; ++i)
    {
        cin >> A[i];
    }

    // Gọi hàm count_dm để đếm ước và bội
    int l, r, d;
    int cnt ;
    for (int i = 0; i < q; ++i)
    {
        cin >> l >> r >> d;

        cnt = count_dm(l, r, d);
        result.push_back(cnt);
    }
}


void output()
{
    freopen(output_file, "w", stdout);

    int s = result.size();

    for (int i = 0; i < s; ++i)
    {
        cout << result[i];
        if (i < s - 1) cout << ' ';
    }
}


int main()
{
    input();
    output();

    return 0;
}

File: hsg12-tinh/2023-2024-vinhphuc/squery/squery.py
import os
import sys

input_file = os.path.join(sys.path[0], 'squery.inp')
output_file = os.path.join(sys.path[0], 'squery.out')

n, q = 0, 0
A = []
result = []

# Hàm dùng để đếm số phần tử là ước hoặc bội của d
def count_dm(l, r, d):
    cnt = 0
    
    # Duyệt mảng A từ vị trí l đến vị trí r
    for i in range(l, r + 1):
        # Xét A[i] là ước hoặc bội của d
        if A[i] % d == 0 or d % A[i] == 0:
            cnt += 1
    return cnt


def input_data():
    global n, q, A, result

    with open(input_file, 'r') as f:
        n, q = map(int, f.readline().split())

        # Cấp phát bộ nhớ cho n + 1 phần tử (vì mảng A đánh chỉ số từ 1)
        A = [0] * (n + 1)

        A[1:] = list(map(int, f.readline().split()))

        # Gọi hàm count_dm để đếm ước và bội
        for _ in range(q):
            l, r, d = map(int, f.readline().split())
            cnt = count_dm(l, r, d)
            result.append(cnt)


def output():
    with open(output_file, 'w') as f:
        f.write(' '.join(map(str, result)))


if __name__ == '__main__':
    input_data()
    output()
